Feature/c code generation tagged unions

chrisdutz opened the pull request:
This is the new approach using tagged unions instead of separate types for every sub-type.

chrisdutz committed:
- Started working on the parser and serializer code ...

chrisdutz committed:
- Major cleanup after getting the compiler to work in CLion

chrisdutz committed:
- Continued implementing the templates for gnerating C parsers/serializers

chrisdutz committed:
Merge branches 'develop' and 'feature/c-code-generation' of https://gitbox.apache.org/repos/asf/plc4x into feature/c-code-generation

chrisdutz committed:
- Switched to union structs for discriminated types

chrisdutz committed:
- Finished implementing the handling of discriminated values in expressions

chrisdutz committed:
- Greatly refactored the C template helper and moved a lot of functionality into a new BaseFreemarkerLanguageTemplateHelper
- Finished support of array fields
- Finished supporting static function calls in expressions
- Finished supporting enum properties in expressions

chrisdutz committed:
- Backported the major changes in the C code-generation back to the java version
- Introduced a BaseFreemarkerLanguageTemplateHelper which contains methods usable for multiple code generations
- Updated the mspec definitions to now use the discriminators directly

chrisdutz committed:
- Continued working on the C code generation
-- Fixed a lot of little issues.
-- Cleaned up the usage of pointers and no-pointers

ottobackwards commented on build-utils/language-c/src/main/java/org/apache/plc4x/language/c/CLanguageTemplateHelper.java:
```diff
@@ -436,8 +476,45 @@ private String toExpression(Field field, Term term, Function<Term, String> varia
 
-    public String toVariableParseExpression(Field field, Term term, Argument[] parserArguments) {
+    public String toVariableParseExpression(ComplexTypeDefinition baseType, Field field, Term term, Argument[] parserArguments) {
         VariableLiteral vl = (VariableLiteral) term;
```
magic string

chrisdutz replied:
??

ottobackwards replied:
"CAST" could be a constant.  If there are other things you could do that too.

chrisdutz replied:
Yeah ... correct ... but I think there's only one place where it's used ...

ottobackwards commented on sandbox/plc4c/generated-sources/modbus/src/modbus_constants.c:
```diff
@@ -24,18 +24,18 @@
 
 // Parse function.
-plc4c_return_code plc4c_modbus_read_write_modbus_constants_parse(plc4c_spi_read_buffer* buf, plc4c_modbus_read_write_modbus_constants** message) {
+plc4c_return_code plc4c_modbus_read_write_modbus_constants_parse(plc4c_spi_read_buffer* buf, plc4c_modbus_read_write_modbus_constants** _message) {
```
to me, when I see _var_name I think of a private var, not only in python but in some c-ish languages I've worked in.  is there some standard that you are writing two using _ in a parameter name?

chrisdutz replied:
Well I had to somehow make sure the names a user can choose in mspec don't interfere with those used internally ... so I prefixed them with a "_" if you have another suggestion, I'm  fine with changing.

ottobackwards commented:
that is too much to look through all of it.
It looks good.  A couple of things that come to mind:

- can we run clang-format after generation?
- should we be generating some logging?
- _message as a parameter is strange to _me_
- so many (*_message)-> everywhere, can we use a typedef or something?

None of these things are for this PR

chrisdutz committed:
- Refactored the code generation to use pointers for complex types and arrays/lists
- Worked on the expression builders to correctly access indexed items of lists

chrisdutz committed:
- Fixed the last syntax-error issue (for now) in the code-generation of the parser code.

chrisdutz committed:
- Added checks for the return codes of parsed sub-types
- Added NULL-checks after the malloc of complex types

chrisdutz committed:
- Fixed a big in the generation of enum property generation

chrisdutz commented:
> that is too much to look through all of it.
> It looks good. A couple of things that come to mind:
> 
> * can we run clang-format after generation?
> * should we be generating some logging?
> * _message as a parameter is strange to _me_
> * so many (*_message)-> everywhere, can we use a typedef or something?
> 
> None of these things are for this PR

Well I would rather have the code generated in a way that it doesn't require formatting ...

I didn't generate any logging as I didn't know of any logging framework to use ... we could whip up a skeleton of functions that we can use for logging and plug in a "printf" logger or whatever we like to do the actual logging ... that might be a good idea.

As I mentioned before ... I wanted to ensure the internal variable names don't clash with the ones a user can use in mspec ... people tent to get anoyed if they have to use other named due to internal stuff.

Why do we need to typedef that ... It's not that anyone is ever going to write that code ... but if you like, of course that's probably a really simple change.

chrisdutz committed:
- Added some indentation to the generation of fields inside discriminated types

ottobackwards commented:
@chrisdutz that is fair enough.  no reason not to keep it like it is

chrisdutz committed:
- Started implementing the different types of serializers for the fields (WIP)

chrisdutz committed:
- Implemented more field types for the serialization

chrisdutz committed:
- Finished implementing the implicit fields
- Fixed the generation of the padding code (which used the "serializer"-arg "lastItem")
- Added skeletons for the lengthInBytes code.

chrisdutz committed:
- Finished implementing the "get_lengh_in_bits" functionality
- Made the switch type also save a reference to the enum constant for that type
